### PR TITLE
Fixes some widgets(`ListView.builder`, `GridView.builder` etc.) state-lose issue

### LINKED ISSUE #100451: States rebuilt when index changed and state is different in ListView.builder (ListView.custom)
## Steps to Reproduce

1. Run the sample code
2. Click on the floating button as many times as you want and watch the logs
3. See that all items are disposed (in the logs) and recreated after each insert

**Expected results:**

The widget state should be kept even when changing index (for different states). 

**Actual results:** 

The widgets should not be disposed after inserting new items and changing the "states rank".  

This is very hard to implement a usual chat app with this issue as we have many `StatefulWidgets` of different kind and we must have a `reversed` and ordered list as you probably see in your daily apps.  

This issue reloads all of the widgets where the state is changing for a given index.  

If I am doing a wrong implementation please feel free to let me know.  
I have also checked `ListView.custom` and observed the same behavior (despite it was suggested in other similar issues). 

<details>
<summary>Code sample</summary>

### `ListView.builder` : 

```dart

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: ListDisposeTest()
      ),
    );
  }
}

class ListDisposeTest extends StatefulWidget {
  const ListDisposeTest({Key? key}) : super(key: key);

  @override
  State<ListDisposeTest> createState() => _ListDisposeTestState();
}

class _ListDisposeTestState extends State<ListDisposeTest> {
  List<int> numberList = [];

  int number = -1;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButton: FloatingActionButton(
       child: Icon(
         Icons.add
        ),
        onPressed: () {
          setState(() {
            number += 1;
            numberList.insert(0, number);
          });
        },
      ),
      body: Stack(
        children: [
          SafeArea(
            child: ListView.builder(
                itemBuilder: (context, index) {
                  return numberList[index] % 2 == 0
                      ? ItemListDisposePair(
                          index: numberList[index],
                          key: ValueKey<int>(numberList[index]))
                      : ItemListDispose(
                          index: numberList[index],
                          key: ValueKey<int>(numberList[index]));
                },
                reverse: true,
                itemCount: numberList.length),
          ),
        ],
      ),
    );
  }
}

class ItemListDispose extends StatefulWidget {
  final int index;

  const ItemListDispose({Key? key, required this.index}) : super(key: key);

  @override
  State<ItemListDispose> createState() => _ItemListDisposeState();
}

class _ItemListDisposeState extends State<ItemListDispose> {
  @override
  void initState() {
    print("init item : ${widget.index}");
    super.initState();
  }

  @override
  void dispose() {
    print("dispose item : ${widget.index}");
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.red,
      margin: EdgeInsets.only(bottom: 20),
      height: 50,
      child: Text(
        "${widget.index} + impair",
        style: TextStyle(color: Colors.white, fontSize: 20),
      ),
    );
  }
}

class ItemListDisposePair extends StatefulWidget {
  final int index;

  const ItemListDisposePair({Key? key, required this.index}) : super(key: key);

  @override
  State<ItemListDisposePair> createState() => _ItemListDisposePairState();
}

class _ItemListDisposePairState extends State<ItemListDisposePair> {
  @override
  void initState() {
    print("init item : ${widget.index}");
    super.initState();
  }

  @override
  void dispose() {
    print("dispose item : ${widget.index}");
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.blue,
      margin: EdgeInsets.only(bottom: 20),
      height: 50,
      child: Text(
        "${widget.index} + pair ",
        style: TextStyle(color: Colors.white, fontSize: 20),
      ),
    );
  }
}
```

</details>

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel stable, 2.10.3, on macOS 12.3 21E230 darwin-x64, locale fr-FR)
    • Flutter version 2.10.3 at /Users/foxtom/Desktop/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7e9793dee1 (3 weeks ago), 2022-03-02 11:23:12 -0600
    • Engine revision bd539267b4
    • Dart version 2.16.1
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/foxtom/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[✓] VS Code (version 1.65.2)
    • VS Code at /Users/foxtom/Desktop/Visual Studio Code.app/Contents
    • Flutter extension version 3.36.0

[✓] Connected device (3 available)
    • moto g 8 power (mobile) • ZY22BNDW2C • android-arm64  • Android 11 (API 30)
    • macOS (desktop)         • macos      • darwin-x64     • macOS 12.3 21E230 darwin-x64
    • Chrome (web)            • chrome     • web-javascript • Google Chrome 99.0.4844.83

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes #100451

`SliverChildBuilderDelegate` without `findChildIndexCallback` may result in state-loss

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/100547

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/100547

File: packages/flutter/test/widgets/list_view_test.dart [hunk 1]
```diff
@@ -76,6 +76,78 @@ class _StatefulListViewState extends State<_StatefulListView> {
 }
 
 void main() {
+  // Regression test for https://github.com/flutter/flutter/issues/100451
```
Comment (chunhtai) [inline packages/flutter/test/widgets/list_view_test.dart:79]: Also add this to all the test

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 2]
```diff
@@ -669,6 +669,9 @@ class PageView extends StatefulWidget {
   /// you are planning to change child order at a later time, consider using
   /// [PageView] or [PageView.custom].
   ///
+  /// The `findChildIndexCallback` argument corresponds to the
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/page_view.dart:672]: explain a bit more why user may want to provide this?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/page_view.dart:672]: Also make this into a template that can be share across all the relevant class

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 3]
```diff
@@ -665,9 +665,14 @@ class PageView extends StatefulWidget {
   /// [itemBuilder] will be called only with indices greater than or equal to
   /// zero and less than [itemCount].
   ///
-  /// [PageView.builder] by default does not support child reordering. If
-  /// you are planning to change child order at a later time, consider using
-  /// [PageView] or [PageView.custom].
+  /// {@template flutter.widgets.PageView.findChildIndexCallback}
+  /// [findChildIndexCallback] corresponds to the
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/page_view.dart:669]:
```suggestion
  /// The [findChildIndexCallback] corresponds to the
```

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 4]
```diff
@@ -194,6 +194,10 @@ abstract class SliverChildDelegate {
   /// This will be called during `performRebuild` in [SliverMultiBoxAdaptorElement]
   /// to check if a child has moved to a different position. It should return the
   /// index of the child element with associated key, null if not found.
+  ///
+  /// If not provided, a child widget may not map to its existing [RenderObject]
+  /// when the order in which children are returned from the children builder changes.
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/sliver.dart:199]:
```suggestion
  /// when the order of children returned from the children builder changes.
```

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 5]
```diff
@@ -665,9 +665,14 @@ class PageView extends StatefulWidget {
...
+  /// [SliverChildBuilderDelegate.findChildIndexCallback] property. If null,
+  /// a child widget may not map to its existing [RenderObject] when the order
+  /// in which children are returned from the children builder changes.
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/page_view.dart:672]:
```suggestion
  /// of children returned from the children builder changes.
```

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 6]
```diff
@@ -429,14 +433,16 @@ class SliverChildBuilderDelegate extends SliverChildDelegate {
   /// Defaults to providing an index for each widget.
   final SemanticIndexCallback semanticIndexCallback;
 
+  /// {@template flutter.widgets.SliverChildBuilderDelegate.findChildIndexCallback}
   /// Called to find the new index of a child based on its key in case of reordering.
   ///
   /// If not provided, a child widget may not map to its existing [RenderObject]
-  /// when the order in which children are returned from [builder] changes.
+  /// when the order in which children are returned from the children builder changes.
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/sliver.dart:440]:
```suggestion
  /// when the order of children returned from the children builder changes.
```

**Review by chunhtai — APPROVED** [review]
LGTM, except some minor nits

**Pull request merged**